### PR TITLE
Enable `if` and `match` in constants behind a feature flag

### PR DESCRIPTION
**Pull request by ecstatic-morse**
This PR is an initial implementation of #49146. It introduces a `const_if_match` feature flag and does the following if it is enabled:
- Allows `Downcast` projections, `SwitchInt` terminators and `FakeRead`s for matched places through the MIR const-checker.
- Allows `if` and `match` expressions through the HIR const-checker.
- Stops converting `&&` to `&` and `||` to `|` in `const` and `static` items.

As a result, the following operations are now allowed in a const context behind the feature flag:
- `if` and `match`
- short circuiting logic operators (`&&` and `||`)
- the `assert` and `debug_assert` macros (if the `const_panic` feature flag is also enabled)

However, the following operations remain forbidden:
- `while`, `loop` and `for` (see #52000)
- the `?` operator (calls `From::from` on its error variant)
- the `assert_eq` and `assert_ne` macros, along with their `debug` variants (calls `fmt::Debug`)

This PR is possible now that we use dataflow for const qualification (see #64470 and #66385).

r? @oli-obk 
cc @rust-lang/wg-const-eval @eddyb

### CODE REVIEW

File: src/librustc_mir/transform/qualify_min_const_fn.rs
Comment (Centril) [inline src/librustc_mir/transform/qualify_min_const_fn.rs]: @oli-obk Also if you remember, what was the reason for disallowing `::Abort | ::Unreachable`?

File: src/test/ui/consts/const-match-pattern-arm.stderr
Comment (Centril) [inline src/test/ui/consts/const-match-pattern-arm.stderr]: Can we gather all control flow tests into `consts/control-flow` so that it's better organized?

File: src/test/ui/consts/control-flow/basics.rs
Comment (Centril) [inline src/test/ui/consts/control-flow/basics.rs]: More test ideas (for other files -- and in other follow up PRs if necessary):

- Test all the various https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/enum.PatKind.html forms and some random walk of that. (No reason to believe it's different, but sometimes that's wrong; well except `Box` patterns).

- Also test `if let` (good idea to not assume a HIR->MIR desugar since that might change, and probably will...).

File: src/test/ui/consts/control-flow/drop.rs
Comment (Centril) [inline src/test/ui/consts/control-flow/drop.rs]: I would prefer if we could separate the good/bad parts of the test into different files (general note).

File: src/test/ui/consts/control-flow/drop.rs
Comment (Centril) [inline src/test/ui/consts/control-flow/drop.rs]: More interesting cases:
```rust
match () {
    () => x,
}
```
and:
```rust
match x {
    Ok(x) | Err(x) => Ok(x),
}
```

File: src/test/ui/consts/control-flow/feature-gate-const-if-match.rs
Comment (Centril) [inline src/test/ui/consts/control-flow/feature-gate-const-if-match.rs]: Would prefer splitting the file and having one be `// check-pass` so that stabilization becomes easier later on.

File: src/test/ui/consts/control-flow/interior-mutability.rs
Comment (Centril) [inline src/test/ui/consts/control-flow/interior-mutability.rs]: Would be good to repeat this for `if let` and `match` (again to not presume the MIR lowering).

File: src/test/ui/consts/control-flow/short-circuit-let.rs
Comment (Centril) [inline src/test/ui/consts/control-flow/short-circuit-let.rs]: Let's also test the values at compile time & run-time.

Would be also good to have e.g.
```rust
const fn foo() -> bool {
    let x = true && return false; // and similar for `||`
    x
}
```

File: src/test/ui/consts/miri_unleashed/enum_discriminants.rs
Comment (Centril) [inline src/test/ui/consts/miri_unleashed/enum_discriminants.rs]: What happens if you remove -Zunleash with the new feature enabled?

File: src/test/ui/consts/single_variant_match_ice.stderr
Comment (Centril) [inline src/test/ui/consts/single_variant_match_ice.stderr]: Seems like this can be `// check-pass` now?

File: src/test/ui/issues/issue-46843.stderr
Comment (Centril) [inline src/test/ui/issues/issue-46843.stderr]: Let's enable the gate here (conditionally) and move the test into `control-flow/` ?

File: src/test/ui/issues/issue-50577.stderr
Comment (Centril) [inline src/test/ui/issues/issue-50577.stderr]: Would be good to conditionally switch on the gate here also.

File: src/tools/compiletest/src/errors.rs
Comment (Centril) [inline src/tools/compiletest/src/errors.rs]: Would prefer if we did the changes to compiletest in a different PR including with tests independent of the added feature gate to ensure compiletest is working properly.

**Comment by Centril** [comment]
Added the feature gate label. @ecstatic-morse Could you please mark relevant past PRs that lead up to this one with that label as well? (It's very helpful for tracking implementation history for stabilization report purposes.)

File: src/test/ui/consts/control-flow/interior-mutability.rs
Comment (oli-obk) [inline src/test/ui/consts/control-flow/interior-mutability.rs]: Do we have a test for

```
const X: Option<Cell<i32>> = {
    let mut x = Some(Cell::new(4));
    if true {
        x = None;
    }
    x
};
```

?

File: src/librustc_passes/check_const.rs
Comment (oli-obk) [inline src/librustc_passes/check_const.rs]: there's a specific macro for generating feature gate errors. I think this should be used here instead of rerolling the logic

File: src/test/ui/consts/control-flow/drop.rs
Comment (oli-obk) [inline src/test/ui/consts/control-flow/drop.rs]: do you want to open an issue about it or are you addressing it already?

**Review by oli-obk — CHANGES_REQUESTED** [review]
I love how simple this is to do after all the refactorings and cleanups you did. Except for a diagnostic nit and some questions this looks ready to go to me

File: src/librustc_passes/check_const.rs
Comment (Centril) [reply src/librustc_passes/check_const.rs]: See https://github.com/rust-lang/rust/pull/66507#discussion_r347256214 .

File: src/librustc_mir/transform/qualify_min_const_fn.rs
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: yea we should probably match on these exhaustively

maybe a single arm for the `FakeRead` and then take its fields apart

File: src/librustc_mir/transform/qualify_min_const_fn.rs
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: probably because we have no const fn code actually hitting it and thus there being no way to test it

Comment (Centril) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: Can't we use `-Z panic=abort`?

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: that just removes the landing pads, but still goes through the regular panic invocation I think. It just doesn't unwind in the end.

Comment (matthewjasper) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: Abort terminators are generated on `#[unwind(aborts)]` functions when compiling with `-C panic=unwind`, which I believe can happen for `const fn`.

File: src/librustc_passes/check_const.rs
Comment (ecstatic-morse) [reply src/librustc_passes/check_const.rs]: My bad. I didn't know that existed.

**Comment by jyn514** [comment]
@ecstatic-morse this is really exciting! I still want to contribute tests but my laptop is currently broken so I'll be helping only sporadically.

Where would I start? I see you added tests in `src/test/ui/consts/control-flow/basics.rs` so I can follow that format, but what sort of things need to be tested? I would mostly be using this for enum matching (things like https://github.com/bytecodealliance/target-lexicon/blob/master/src/targets.rs#L126) but I imagine that isn't the hard part of this feature.

File: src/test/ui/consts/miri_unleashed/enum_discriminants.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/miri_unleashed/enum_discriminants.rs]: Good point. This should no longer require the flag.

File: src/librustc_passes/check_const.rs
Comment (ecstatic-morse) [reply src/librustc_passes/check_const.rs]: This is mutually exclusive with the change below (`while let` => `while`) unless we want to change `LoopSource::name`, which is probably fine.

Comment (Centril) [reply src/librustc_passes/check_const.rs]: I would prefer for this one to take priority and perhaps think about the output below later ; my main point below was just that we should be consistent wrt if and while :)

File: src/test/ui/consts/control-flow/drop.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/drop.rs]: We reject this for the same reason we reject:

```rust
const fn foo() -> Vec<i32> {
    let vec_and_bool = (Vec::new(), true);
    vec_and_bool.0
}
```

My thought was that we could special case that a move out of the only field of a struct or enum variant also clears `Drop` for the container. We may want to defer this to a follow-up PR? I'm not quite sure what all the implications of this are.

Comment (oli-obk) [reply src/test/ui/consts/control-flow/drop.rs]: Oh yea, please defer. Being overly restrictive at the start is totally the right way

File: src/tools/compiletest/src/errors.rs
Comment (ecstatic-morse) [reply src/tools/compiletest/src/errors.rs]: See #66524

File: src/test/ui/consts/const-match-pattern-arm.stderr
Comment (ecstatic-morse) [reply src/test/ui/consts/const-match-pattern-arm.stderr]: I've done this for the obvious files, and will try to do this from now on.

File: src/test/ui/consts/control-flow/drop.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/drop.rs]: It's now done for this test and I'll do this in the future.

File: src/test/ui/consts/control-flow/interior-mutability.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/interior-mutability.rs]: Added [here](https://github.com/rust-lang/rust/pull/66507/files#diff-e2a59c2bcf3c9ce2dd92a137f9999680R16-R22).

File: src/test/ui/consts/control-flow/drop.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/drop.rs]: I added the reduced version as a failing test case.

**Comment by ecstatic-morse** [comment]
I've rebased with @Centril's changes to the HIR const-checker applied. I now use `emit_feature_err` instead of custom logic. I also moved the `compiletest` changes into their own PR (#66524). I squashed all the test changes into one commit since the error messages are in flux and rebasing after each `bless` became a pain.

File: src/test/ui/consts/control-flow/interior-mutability.rs
Comment (Centril) [reply src/test/ui/consts/control-flow/interior-mutability.rs]: Did you miss this one?

Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/interior-mutability.rs]: If I marked it resolved, it was by accident. I don't think it adds anything to repeat each qualifs test three times, however.

File: src/test/ui/consts/control-flow/feature-gate-const-if-match.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/feature-gate-const-if-match.rs]: I prefer the `// revisions:` approach to having two copies of the file while the feature flag exists.

File: src/test/ui/consts/control-flow/interior-mutability.rs
Comment (oli-obk) [reply src/test/ui/consts/control-flow/interior-mutability.rs]: I agree. If the lowerings start differing in ways that are problematic here we're gonna see other problems, too.

File: src/librustc_mir/transform/qualify_min_const_fn.rs
Comment (ecstatic-morse) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: I think you two are discussing different things here? It seems like we should indeed check the LHS of `SetDiscriminant`, although we need to do this in `check_consts` as well, and I think it's unrelated to this PR. We should be using a `Visitor` here and just calling `super_statement`, but this is blocked on #65572.

`FakeRead` is handled properly at the moment IMO.

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: Yea I didn't read the patterns correctly. Ignore that comment

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: Opened https://github.com/rust-lang/rust/issues/66556

File: src/librustc_mir/transform/qualify_min_const_fn.rs
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: I opened https://github.com/rust-lang/rust/issues/66557

File: src/test/ui/consts/control-flow/basics.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/basics.rs]: @jyn514, this would good be a good place to contribute if you're still motivated. "Interesting" cases could be added to `basics.rs` (or another file) that use more types of patterns in `match` and `if let`. If you push to a branch I can cherry pick them into this PR (keeping you as the author of course), or you can open a separate PR once this is merged.

**Comment by oli-obk** [comment]
Does this PR still count as WIP? Seems ready to me

File: src/test/ui/consts/control-flow/feature-gate-const-if-match.rs
Comment (Centril) [reply src/test/ui/consts/control-flow/feature-gate-const-if-match.rs]: @ecstatic-morse In that case, please leave a note so that the file isn't removed whole-sale when we want to stabilize and so that it is converted into `// check-pass`.

**Comment by ecstatic-morse** [comment]
@oli-obk Yeah, I think this is ready to go. I'd like to have a few more tests for interior mutability/`Drop`, but it's quite hard to come up with novel ones. Hopefully nightly users are more creative :smile:.

**Comment by ecstatic-morse** [comment]
Oh, this will have to be merged after #66524. They shouldn't conflict though.

**Comment by Centril** [comment]
@ecstatic-morse would be good to address https://github.com/rust-lang/rust/pull/66507#discussion_r348463724 first tho.

File: src/doc/unstable-book/src/language-features/const-if-match.md
Comment (goto-bus-stop) [inline src/doc/unstable-book/src/language-features/const-if-match.md]:
```suggestion
[#49146]: https://github.com/rust-lang/rust/issues/49146
```

File: src/libsyntax/feature_gate/active.rs
Comment (Centril) [inline src/libsyntax/feature_gate/active.rs]: Btw: a useful trick for the future, if you want to avoid rebasing due to this, is to move this before `never_type_fallback`.

File: src/test/ui/consts/control-flow/feature-gate-const-if-match.rs
Comment (ecstatic-morse) [reply src/test/ui/consts/control-flow/feature-gate-const-if-match.rs]: There's a comment at the top of the file now.

**Comment by ecstatic-morse** [comment]
@oli-obk Can you do a final review of this when you have the time?

**Pull request merged**
